### ISSUE TITLE
Need help deciding on Better Handling of Object Types

### ISSUE THREAD

**Issue by mohpor** [CosmicMind/Graph#97]
Hi again,

Graph does a great job handling everything I need, except, I'm a closed minded developer who needs every object to be immaculate. Anyways, Referring every object I'm working with as `Entity` and pass them around is something I can't get my head around it. Maybe I'm seeing it from a wrong perspective (it's highly likely), but I cannot find a good resource on it.

I tried to overcome this dilemma by introducing custom types based on `Entity` and simply add properties based on my needs to the `Entity` based objects. Here is an example:

``` swift

class MyObject: Entity {

  private struct FieldNames {
  static let EntityType = "MyObject"
    struct Properties {
      static let Field1Name = "field1"
      static let Field2Name = "field2"
    }
  }

  var Field1: String? {
    return self[FieldNames.Properties.Field1Name] as? String
  }

  var Field2: String? {
    return self[FieldNames.Properties.Field2Name] as? String
  }
}

```

This way I can work with entities in a more civilized manner. I have actually a more complex set of properties and they synchronize with graph with ease but I don't want to get into that for now.

Now the problem: Because Graph returns `Entity` objects and they are initialized as `Entity` I can't downcast them to my own types. It seems obvious, I know. Extending Entity is not an option here because it affects all `Entity`objects not just my types.

Needless to say, all these goes for `Relationship`s too.

Questions:
1- Is this a good pattern at all? Am I wrong thinking about such solutions and should swallow the pill and work with graphs as they are intended to work? I'm still thinking in pure CoreData mindset which I had control over my object types. (Pattern - AntiPattern)

2- How can I overcome this dilemma? I was under the impression that I can make custom type converters but apparently I was wrong. How do you think I can make this work?

3- Is it a good idea to design the object as a composite (I'm not even sure I'm using the right pattern name here) and use the entity as an internal mechanism to return values based on that? Something like this:

``` swift
class MyObject {
  private struct FieldNames {
  static let EntityType = "MyObject"
    struct Properties {
      static let Field1Name = "field1"
      static let Field2Name = "field2"
    }
  }
  let entity: Entity!
  init?(entity: Entity){    
    self.entity = entity
    guard entity.type == FieldNames.EntityType else {
      return nil
    }
  }

  var Field1: String? {
    return entity[FieldNames.Properties.Field1Name] as? String
  }

  var Field2: String? {
    return entity[FieldNames.Properties.Field2Name] as? String
  }

}
```

Is there any caveat in doing so? Memory leaks maybe? Synchronization?

**Comment by daniel-jonathan** [comment]
Hi @mohpor, 

Really cool that you are experimenting with Graph :) It's also interesting to read about your approach when using it. 

Graph is meant to be used in the way that you think. You can make a Wrapper Interface around the Entity instances, which you referred to as the Composite Pattern, or you can use as is. I would recommend a hybrid approach. You may want to make a Factory class for creating objects and initializing them. If you did that, wrapping objects would be self contained and easy to maintain. For example:

``` swift
public struct User {
    private let entity: Entity

    /// FirstName property.
    public var firstName: String? {
        get {
            return entity["firstName"] as? String
        }
        set(value) {
            entity["firstName"] = value
        }
    }
}

public struct UserFactory {
    static private let graph: Graph = Graph()

    /// Create a new User.
    static public func createInstance() -> User {
        return User(entity: Entity(type: "User"))
    }

    /// Fetch all the Users.
    static public func collection() -> Array<User> {
        var users: Array<User> = Array<User>()
        for entity in graph.searchForEntity(types: ["User"]) {
            users.append(User(entity: entity))
        }
        return users
    }
}
```

Then in your application you could do this:

``` swift
let graph: Graph = Graph()

var user: User = UserFactory.createInstance()

print(user.firstName)

user.firstName = "Daniel"

print(user.firstName)

graph.save()

let users: Array<User> = UserFactory.collection()

print(users)
```

I generally use Graph as is, and when I need a more concrete API for a team of developers let's say, where they do not all know Graph, the Hybrid approach is excellent. 

Thoughts?

**Comment by mohpor** [comment]
@danieldahan , Hi,

I Like the hybrid approach, it is practical, but, I hoped there was a way that we didn't have to wrap all functionalities as required and remained a first class citizen in `Graph`, you know what I mean? This way, we're gonna have to adapt to new `Graph` API everytime something changes profoundly and the worst part is, potentially in many class implementations.

Now, how about letting `Graph` services be generic (types should be sublasses of `Entity`) and help implementing such patterns with more ease?
For example: search method could be (pseudo code): `searchForEntities<T: Entity>....` 
How about that?

**Comment by daniel-jonathan** [comment]
Hmmm... I think it would be neat to have an approach to helping Graph be used in the method you are looking for. As mentioned, I use it as is, as I feel it is generic. You pass in types to the Classes that exist in Graph. 

Over the month of March, we had planned to really take Graph to the next level. Your suggestion is a good addition to our considerations :)

**Comment by mohpor** [comment]
I can fashion a PR if it is alright with you guys.

**Comment by daniel-jonathan** [comment]
Yeah, that would be great. Let's leave this thread open to keep the discussion going.

**Comment by mohpor** [comment]
Agreed, I'm on it.

**Comment by daniel-jonathan** [comment]
Great. Thank you, and please send the PR to the development branch.

**Comment by mohpor** [comment]
Of course.

**Comment by mohpor** [comment]
Ok, I need help.
In my quest to do my assignment, I'm facing a concrete wall .

![screen-shot-2016-03-07-at-9 41 58-am](https://cloud.githubusercontent.com/assets/4954588/13562051/0622c3fe-e449-11e5-83d0-97ac4c2496cc.png)

### tl;dr:

To be able to initialize a generic object (subclasses of another object) I need to have `required init` constructors (Swift imposes this). And to make the `init(object: ManagedEntity)` class `required`, it needs to have the same scope as the class itself, obviously. To make it `public` I need `ManagedEntity` to be `public` not `internal` and it's a downward spiral from there.

Does anyone know a better solution?

**Comment by daniel-jonathan** [comment]
Can you explain a little more of what you are doing here and then I can help with a solution?

**Comment by mohpor** [comment]
I'm trying to instantiate an `Entity` using a generic type (subclasses of `Entity`) because in more than one places we need to insatiate Entities and `init(object: ManagedEntity)` is not going to work for us. For instance at the end of the `searchForEntity(types: groups: properties)` we need to instantiate Entities using either a `ManagedEntity` or a `ManagedObjectID`.

**Comment by daniel-jonathan** [comment]
So you should never create an Entity using the internal objects. That would create issues. You could do this. In the Entity class, add the required attribute to the initializer:

``` swift
public required convenience init(type: String) {
    self.init(object: ManagedEntity(type: type))
}
```

Then you can do this: 

``` swift
struct GraphFactory {
    static func createInstance<T: Entity>(type: String) -> T {
        return T(type: type)
    }
}
```

**Comment by daniel-jonathan** [comment]
@mohpor How is this coming along?

**Comment by mohpor** [comment]
@danieldahan, not so good,
I was busy the past 3 days and could find any time to work on it.

You suggestion was my main approach but it is not enough, because we are not always initiating a new object, sometimes we are initializing it using a `ManagedEntity` for example in search operations, we already have an object, we are just creating an object based on other objects. good thing is, we still can send around the `ManagedObjectID` right? I ca rewrite those methods and introduce a required initializer that accepts `ManagedObjectID`s and create `Entity` subclasses from them. Do you think it's a good idea? Any known caveat? Performance loss?

**Comment by mohpor** [comment]
Well, I managed to wrap my head around a practical way to do it.
I'm guessing I've broken a dozen design patterns here but it is very minimal so to speak.
Swift 1 had a very useful function named `__conversion` in which you could do this flawlessly. I wish it was still possible to use it.
Anyway, it passed the test and I hope you're happy with it.

**Comment by daniel-jonathan** [comment]
Generally, one should never actually interface with the NSManagedObject, externally that is from Graph, as well, it is not a good idea to provide any direct access to the NSManagedObject, as it creates exposure and thus issues. I think I am losing insight into what we are accomplishing here. 

Can you write a Pseudocode example of what would really be the gain in the code changes? Thank you!

**Comment by mohpor** [comment]
@danieldahan, You're right.
It seems unattainable at the moment.
I'm going to implement a Decorator pattern for my needs.
I think it's safe to close this issue and reject the pull request.
thanks
M.

**Comment by daniel-jonathan** [comment]
Great! Thank you for the effort :)
